src/adapter/search-response-adapter/pagination-adapter.ts

```diff
@@ -1,21 +1,55 @@
-/**
- * Slice the requested hits based on the pagination position.
- *
- * @param  {Record<string} hits
- * @param  {number} page
- * @param  {number} hitsPerPage
- * @returns {Array}
- */
-export function adaptPagination(
-  hits: Record<string, any>,
-  page: number,
+import type {
+  MeiliSearchResponse,
+  PaginationState,
+  InstantSearchPagination,
+} from '../../types'
+import { ceiledDivision, floorDivision } from '../../utils'
+
+function adaptNbPages(
+  searchResponse: MeiliSearchResponse<Record<string, any>>,
   hitsPerPage: number
-): Array<Record<string, any>> {
-  if (hitsPerPage < 0) {
-    throw new TypeError(
-      'Value too small for "hitsPerPage" parameter, expected integer between 0 and 9223372036854775807'
-    )
+): number {
+  if (searchResponse.totalPages != null) {
+    return searchResponse.totalPages
+  }
+
+  return ceiledDivision(searchResponse.hits.length, hitsPerPage)
+}
+
+function adaptPage(
+  searchResponse: MeiliSearchResponse<Record<string, any>>,
+  hitsPerPage: number
+): number {
+  const { limit, page } = searchResponse
+
+  // Finite pagination environment
+  if (page === 0) {
+    // Should not happen but safeguarding in case it does
+    return 0
+  }
+  if (page != null) {
+    return page - 1
+  }
+
+  // Scroll pagination environment
```

Comment:
Can you use the same terms, like `estimated`?

Reply:
I don't think `estimated` makes it clear :/ Maybe `Lazy pagination` ?

Reply:
I say estimated because of the `estimatedTotalHits` :)

Reply:
We don't use this number in the code base, it does help the user understand for me